gateway/handle_http_auth.go

```diff
@@ -119,6 +119,44 @@ func (gw *Handle) sourceIDAuth(delegate http.HandlerFunc) http.HandlerFunc {
 	}
 }
 
+// authenticateDestinationIdForSource middleware to authenticate destinationId in the X-Rudder-Destination-Id header.
+// If the destinationId is invalid, the request is rejected.
+// destinationID authentication should be performed only after source is authenticated and source is present in context
+// Following validations are performed
+//  1. Destination should be present for source config
+//  2. Destination should be enabled for the source
+func (gw *Handle) authenticateDestinationIdForSource(delegate http.HandlerFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		reqType := r.Context().Value(gwtypes.CtxParamCallType).(string)
+		arctx := r.Context().Value(gwtypes.CtxParamAuthRequestContext).(*gwtypes.AuthRequestContext)
```

Comment:
Can we please avoid the casting repetition we're doing everytime we get/set a context value for these types?

It's risky since it could lead to a panic (in case it is set as a type and read as another). There are [safer ways](https://github.com/rudderlabs/rudder-server/blob/v1.18.2/warehouse/utils/context.go) to do this, I suggest doing the same here.